### ISSUE TITLE
Type checking and API docs with TypeScript JSDoc annotations

### ISSUE THREAD

**Issue by ahocevar** [openlayers/openlayers#8448]
We are currently transitioning the code base from type annotations understood by JSDoc to TypeScript JSDoc annotations. This ticket is an overview on progress and remaining tasks:

- [x] #8345 Rewrite JSDoc annotations to those [understood by TypeScript](https://github.com/Microsoft/TypeScript/wiki/JSDoc-support-in-JavaScript)
- [x] #8449 New tool chain to generate API docs from the source code and JSDoc annotations
- [ ] Investigate TypeScript integration: will `--allowJS` be enough for users, or do we need to publish `.d.ts` files?

Since TypeScript knowledge is limited among the OpenLayers dev community, we appreciate any help on this. Also, please comment if we are missing something in the list above.

**Comment by OSHistory** [comment]
@ahocevar 
Thanks! I'll take the liberty to ping some users that showed interest in the whole 
ts-issue, so we can pool some manpower.

**Comment by OSHistory** [comment]
I pinged you because your activity on the openlayers project indicates you might 
be interested in the current plan to publish typescript definitions within the openlayers 
ol-package (as opposed to use the outdated DefinitlyTyped version, which used to 
worked well for the 4.x series BTW thanks @yairtawil  

If you are in fact not interested in the issue (anymore), I am sorry to have 'spammed' you. 

@8byr0
@achimha
@agxs 
@bjornharrtell
@BobbyT
@BrkCoder
@ca0v
@craigsnyders
@cvenzin
@DarthPinguin
@fgravin
@gberaudo
@h3imdall
@hei-pa
@icholy
@idevtech
@josepgomes
@jumpinjackie
@martin-reiterer
@mrfx
@niftydevelopment
@quentinol
@vrmerlin
@wulfsberg
@yairtawil

**Comment by wulfsberg** [comment]
I have no experience with the JSDoc-annotation style of type definitions for TypeScript, so I'll be of limited use there
I am primarily coming at it as a consumer of OpenLayers in Angular applications, and have never been deep into the build stack of OpenLayers itself, so I'll probably be most useful as a guinea pig for the integration.

**Comment by icholy** [comment]
Once the transition is complete, the typescript compiler can automatically generate a .d.ts file from the sources. Shipping with the definition file makes it easier to consume.

**Comment by marcjansen** [comment]
I wanna also mention

@chrismayer
@kaivolland
@dnlkoch
@jansule
@weskamm
@atd-schubert
@annarieger
and
@ahennr

... if you can help, this would be much appreciated. If you cannot help or don't want to, this is perfectly fine and I am sorry for spamming you.

**Comment by ahocevar** [comment]
If we need to generate `*.d.ts` files, the question is how to generate them. Looks like `tsc` won't be of much help, see Microsoft/TypeScript#21455.

**Comment by icholy** [comment]
@ahocevar that's very unfortunate. We might be able to leverage https://github.com/jumpinjackie/jsdoc-typescript-plugin until tsc supports it.

@jumpinjackie

**Comment by bampakoa** [comment]
Hi all,

There is also [tsd-jsdoc](https://github.com/englercj/tsd-jsdoc) as an option. I am currently using it to generate types for [CesiumJS](https://cesiumjs.org/) and integrates very nice for the time being. Currently, we modify the jsdoc block tags of the CesiumJS source so that it helps the `tsd-jsdoc` compiler to produce as much accurate output as possible.

**Comment by josepgomes** [comment]
If I'm not mistaken, DefinitelyTyped used tsd-jsdoc for openlayers 4.

**Comment by jumpinjackie** [comment]
@icholy Sorry, I'm certain you won't get much success with this. That plugin was geared around the OL (pre 5.x) case and not for ES6 modules in general.

**Comment by jumpinjackie** [comment]
I have an idea about how we can generate a mostly accurate `d.ts` for OpenLayers (and as a double-whammy, possibly also solve the new API docs problem in #8449)

Basically, you want to use [TypeDoc ](https://github.com/TypeStrong/typedoc) as the API generation tool and take advantage of the fact that [it can support allowJs mode of TypeScript](https://github.com/TypeStrong/typedoc/issues/643) to support API doc generation from non-TypeScript sources.

The main reason for choosing TypeDoc is that it also supports JSON output (via `--json`) which will give you full definition of the module and its classes/methods/parameters/return types/etc/etc. Then in terms of generating the `d.ts`, it's just reading the data from this JSON file and writing out the appropriate class/module declarations.

This approach is blocked by #8357. If you [remove the offending typedef](https://github.com/openlayers/openlayers/issues/8357#issuecomment-416225381) as suggested by @aikrez in that issue, then the docs and JSON can get generated by TypeDoc.

Attached is what the docs and JSON files produced by TypeDoc look like

[docs.zip](https://github.com/openlayers/openlayers/files/2333268/docs.zip)
[ol.zip](https://github.com/openlayers/openlayers/files/2333269/ol.zip)

I'm sure if you look at the attached JSON file, that it should not be hard to write a simple post-processor tool to generate the required `d.ts` file.

**Comment by ahocevar** [comment]
Thanks for your thoughts @jumpinjackie. Here is an update of what we've achieved so far and what we're working on:

TypeDoc does not have full support for JSDoc `@typedef` and `@enum` annotations yet. I am currently working on that, see my https://github.com/ahocevar/typedoc/tree/typescript-3.1.x branch.

Regarding the offending `@typedef` annotations, we will be transforming them to annotations that are fully understood by TypeScript v3.1.x. See #8345.

Using TypeDoc's JSON output is definitely an option, but looking at Microsoft/TypeScript#21455, I think a fresh start on that pull request and giving TypeScript the capability to create `*.d.ts` files from a JSDoc type annotated JavaScript code base would be a better invested effort, which would also benefit more projects and users.

**Comment by schmidtk** [comment]
@tschaub & @ahocevar - Myself and @wallw-bits would like to help out with the TypeScript effort. We're both new to TypeScript, but have been using OpenLayers extensively for several years in [OpenSphere](https://github.com/ngageoint/opensphere).

It looks like a lot of the remaining errors are related to lack of `@abstract` support, but we can start working off other errors. To avoid duplicating effort with other developers, would you prefer we submit smaller PR's handling a few files at a time (I did this in #8643), or keep a WIP PR open and commit to that as we work (similar to #8345)?

**Comment by tschaub** [comment]
Thanks for helping out @schmidtk.  I think small PRs will probably be easier to manage (compared to a long lived big one).  If there is something that you think will require significant effort, opening a ticket describing the problem and saying you are going to work on it could help reduce the change of duplicate efforts.

**Comment by ahocevar** [comment]
News on API doc generation with TypeDoc: Trying to make TypeDoc work with all JSDoc type annotations is a fruitless effort (see https://github.com/ahocevar/typedoc/tree/typescript-3.1.x). However, TypeDoc's `--includeDeclarations` option will help us produce good docs. The "only" thing missing to get these declarations is `*.d.ts` files.

So the next step will be to make TypeScript's `allowJs` and `declarations` options work together, see Microsoft/TypeScript#7546. I have started working on a pull request for that (Microsoft/TypeScript#27478), but my insights into TypeScript are too limited to finish the effort alone. Anyone able and willing to help please comment on that pull request so we can coordinate.

**Comment by Diaaz** [comment]
Any news on this? I still can't get openlayers 5 working with typescript / intellisense in VSCode. The 'ol' types package is still on 4.6.3.

**Comment by ahocevar** [comment]
Making TypeScript support `*.d.ts` generation from JSDoc annotated JavaScript is an effort too big for the purpose of our library. As an alternative, we'll be publishing two packages in the future: one with the original ES6 sources that support IntelliSense and type checking through `tsc`, and another with ES5 sources for zero config use with bundlers like the current `ol` package.

**Comment by Diaaz** [comment]
Thanks for that. Do you know when that version will be released? Is there any way to use Openlayers v5 with typescript now? The @types/ol package has not been updated to 5.x.

**Comment by ahocevar** [comment]
There is a way. Instead of installing `ol@5.3.0`, you install `git://github.com/openlayers/openlayers.git#v5.3.0`. Then, instead of importing from `ol` (e.g. `import Map from 'ol/Map';`, you import from `ol/src/ol` (e.g. `import Map from 'ol/src/ol/Map';`).

**Comment by gberaudo** [comment]
Npm sometimes have difficulties with git repos. An alternative is to use a tarbal URLl: `https://api.github.com/repos/openlayers/openlayers/tarball/v5.3.0`.

**Comment by gberaudo** [comment]
@Diaaz, do you have
```json
"allowJs": true,         
"checkJs": true, 
```

In your tsconfig.json/compilerOptions?

**Comment by ahocevar** [comment]
Here is a minimal setup that works: https://gist.github.com/ahocevar/9a7253cb4712e8bf38d75d8ac898e36c

**Comment by gimerstedt** [comment]
appreciate the work you're doing!
i can get your minimal setup to actually go to the definitions in vs-code (not sure what it's using behind the scenes - never use it) but when i correct the file ending to ts it'll stop working.
i can't get it working with tsserver from neovim even with allowJs and checkJs using typescript 3.1.6.

we're not really missing any features if sticking to 4.x long term but it would be nice to get performance upgrades and perhaps even tree-shaking in the future.

what's preventing you from generating definitions atm?

**Comment by ahocevar** [comment]
> what's preventing you from generating definitions atm?

Microsoft/TypeScript#7546, i.e. definitions cannot be created with `"allowJs": true`.

**Comment by gimerstedt** [comment]
and just changing the extensions and flipping allowJs to false wouldn't result in actually typed typescript as the jsdocs are ignored?

**Comment by Diaaz** [comment]
Setting allowjs and checkjs does not work, and using the gist from @ahocevar generates tons of errors on build in an existing typescript project, because all js is compiled. Is there any way to update the  @types/ol  package? That worked perfectly for the last year.

**Comment by tschaub** [comment]
The `@types/ol` package has always been maintained by volunteers outside of the OpenLayers project.  If there are any TypeScript users with experience creating these definition files and interest in maintaining them, I'm sure your contributions to the `@types/ol` package would be welcome.  I think this is the project to look into: https://github.com/DefinitelyTyped/DefinitelyTyped

**Comment by gimerstedt** [comment]
@tschaub i do not install deps without types and if there are no alternatives, i wrap them (and type the entry-points).

**Comment by Diaaz** [comment]
@tschaub I only use typed dependencies, the errors come from dependencies from OpenLayers. The proposed solution from @ahocevar includes OpenLayers in the complete build.

**Comment by ahocevar** [comment]
@Diaaz You could make the type issues from OpenLayers dependencies go away by adding these types as `devDependencies` in your package.json:
```
    "@types/arcgis-rest-api": "^10.4.3",
    "@types/geojson": "^7946.0.4",
    "@types/pbf": "^3.0.1",
    "@types/rbush": "^2.0.2",
    "@types/topojson-specification": "^1.0.0",
```

**Comment by ca0v** [comment]
@ahocevar a project template for tsc development that doesn't require rollup/webpack, etc. would really help.

I'm not sure those dependencies you mention are getting transformed into an TSC consumable format.  I know rbush is still using node syntax.  If a pre-processor converted it to ES6 then maybe ol5 users could link using just tsc with "allowJs" but those dependencies don't allow it as-is.  See https://github.com/mourner/rbush/blob/v2.0.2/index.js#L6.

**Comment by ca0v** [comment]
It would be helpful if that gist could be extended to show how to compile using tsc.  Or is the point only to make vscode happy?

TL;DR;
It would be nice to be able to build using *just* tsc.  I looked at the gist and got it to build using `tsc --allowJS ./index.js --outDir ./bin` but it only produces a transpiled version of index.  I modified it to use amd thinking it might concatentate but it didn't.  And if it did it would choke on rbush and pbf.  See  https://github.com/mourner/rbush/blob/v2.0.2/index.js#L6 for an explanation.  Until those get ported or transpiled to ES6 we can't really build using just tsc.

If they were to get transpiled we could build directly to a single amd file with tree-shaking using just tsc but until that time I think rollup/webpack, etc. is still necessary.

**Comment by tschaub** [comment]
> It would be helpful if that gist could be extended to show how to compile using tsc.

There is no need to compile the OpenLayers sources with `tsc` – they are already JavaScript.  `tsc` is only used for type checking (and this can be done with or without vscode).

**Comment by ca0v** [comment]
@tschaub `tsc` can transpile js files.  In an es6 environment `tsc` can build/shake w/out help from rollup/webpack.  One build option I would like to see is the ability to produce a single AMD file using nothing but `tsc` and loading with `requirejs`.  That's a fast development experience.  Couple with `uglifyjs` and it's possible to run compressed code and debug typescript in chrome.

**Comment by tschaub** [comment]
@ca0v – I understand that `tsc` *can* transpile files, but our intention is to use it for type checking.  If others want to use it to do more, that is great.  But we also want the library to work well for people who don't have any interest in TypeScript, so `tsc` won't be deeply integrated or required in any way.

**Comment by ca0v** [comment]
@tschaub - I was under the impression you were migrating toward typescript due to the recent changes.  I'm unclear how tsc is helping you and the gist doesn't clear it up (for me).

**Comment by icholy** [comment]
@ca0v are you talking about the DX of hacking on openlayers or consuming it as a library?

**Comment by OSHistory** [comment]
@ahocevar as @gimerstedt pointed out, the workaround only works for me 
if the file as a *.js extension. Renaming to *.ts kills Intellisense. 

Since I had some forks sitting around,waiting to upgrade, but I did want 
to have some autocompletion, I went ahead and started a small shell/python-based
project to generate typescript-declarations from the jsdoc-annotated source files. 
The declaration files pass `tsc` and I managed to use the typings in a medium sized 
ol-based library and some other smaller projects. The result is far from perfect, 
but could be a starting point for new types on DefinitleyTyped. As far as i understand 
the original openlayers-types where generated with dts-gen and manually reviewed, 
the ol-types basically was a wrapper around the openlayers-types. 

The repository is under: 

https://github.com/OSHistory/generate_ol_types

I uploaded the resulting types for quick review:

[@types.zip](https://github.com/openlayers/openlayers/files/2630357/types.zip)

**Comment by ca0v** [comment]
@icholy main concern is looking for a way to consume with `tsc` alone. 
 I tried it a while back (https://github.com/ca0v/ol4-lab/tree/v4.0.1) but ran into the same issues I'm seeing today (node-style requires).  I mentioned it on https://github.com/openlayers/openlayers/issues/6494.

**Comment by icholy** [comment]
@ca0v I don't think that use-case is valid enough to justify the work/change involved. Let's try to make these changes as incremental as possible.

**Comment by OSHistory** [comment]
> The next release probably won't be out before February, because we're currently working on a major overhaul of the library.

Just wondering if you're talking major, minor or even bug fix. As documented in the  issue
https://github.com/openlayers/openlayers/issues/8815
production builds inside angular fail with the current package, which is fixed by 
adding proj.js sideEffects (merged in master). Adding and (remembering to add) after 
fresh installs is a bit tedious, so a patch with 5.3.1 would be greatly appreciated.

**Comment by JulioJu** [comment]
Hi !

Thanks to use Type checking and API doc with TypeScript ;-)

I nave a question. How to declare a OpenLayers type ? For instance, I have a TypeScript function like

```typescript
import  View from 'ol/View';
const myFunc = (viewInstanceParam: typeof View): void => {
// or const myFunc = (viewInstanceParam: typeof import('ol/View').default): void => {
// here intellisense works, but he believes it's a `Function`
};
const viewInstance = new View();
// here IntelliSense works perfectly, it knows that it's an instance of View. 
// Bellow we have error: "2345 Argument of type 'View' is not assignable to parameter of type 'typeof View' ".
myFunc(viewInstance);
```

Please, have you and Idea how to declare the type view ?

I've also tried with type `ol/View`, `View` and many others without success. I've passed lot of time on this problem, without success.

Please, could you help me ?

Thanks in advance !

**Comment by wulfsberg** [comment]
`(viewInstanceParam: typeof View)` is wrong. You just want `(viewInstanceParam: View)` to declare the type.

It is a fairly bizarre reaction from the compiler, though. `typeof` is an operator which returns a *string* describing the parameter. It is about the same as defining your type to be `(viewInstanceParam: 2+2)`. I honestly have no idea what the TypeScript compiler is attempting to do with it.

**Comment by icholy** [comment]
@wulfsberg when used in the type context, `typeof` can be used to get the type of an expression. It's a typescript feature.

**Comment by wulfsberg** [comment]
Interesting. I've never seen it used that way. Is that to pick out the type signature of anonymous lambdas or similar? (I assume it's https://github.com/microsoft/typescript/issues/6606. Something to read up on :-D )

**Comment by JulioJu** [comment]
Thanks a lot for your answer, but as I said I have already tested  `(viewInstanceParam: View)` without success. But I have error `2304 Cannot find name 'View`. Have you the same issue  on a TypeScript project ?

**Comment by wulfsberg** [comment]
@JulioJu I have stalled hard on the fact that the JSDoc annotations simply cannot be used with a packaged library (per https://github.com/Microsoft/TypeScript/issues/7546) which makes the whole JSDoc maneuver pretty much a non-starter for my current situation, so I have never actually gotten a version of this beta running.

**Comment by JulioJu** [comment]
@wulfsberg 
yes I have already seen Microsoft/TypeScript#7546 ,  but my goal is not generate declaration files from JsDoc (I've already seen that actually it works not very well), but know how to describe type of an instance of an object. 

As I've said before, actually `tsc` know perfectly the type of `const viewInstance = new View();` : intelliSence retrieves perfectly properties of `viewInstance`. 

But my goal is to do the same when we use a function. As it, we must describe the type of `viewInstance`

```typescript
const viewInstance: typeof ???????? = new View()
```

*  As I said before, `typeof View` is not the instance of the object, but only a newable / Constructor Function. 

* As described at https://www.typescriptlang.org/docs/handbook/generics.html
```typescript
import  View from 'ol/View';
const myFunc = (view: new() => typeof View): void => {
// the type of `view' is the Constructor Function 'ol/View.js'
};
const viewInstance = new View();
// here IntelliSense works perfectly, it knows that it's an instance of `ol/View.js`. 
myFunc(viewInstance);
```

* I have tested with `typeof new View()` or `typeof Object.create(View.prototype)` without success.

Maybe it's a limitation of the `typeof` syntax. 

Maybe I could post a new issue on typescript that reference https://github.com/Microsoft/TypeScript/issues/6606 ? 

What do you think about that ? 

P.S. thanks for the link to ng7-ol5, but it seems does not work actually: see https://github.com/OSHistory/ng7-ol5-demo/issues/2

**Comment by wulfsberg** [comment]
@JulioJu Using `typeof` is simply wrong in this case.
It is an operator which gives you the type of an *existing instance* (which, in case of View is a function, since that's how you create classes in the underlying JavaScript).
You say that you have already tried `(viewInstanceParam: View)` without success. I don't know which errors that caused, but those are the ones to chase. It is absolutely the correct notation, and you will never get  `(viewInstanceParam: typeof View)` to do what you want.

**Comment by JulioJu** [comment]
@wulfsberg thanks a lot for your answer. Please see https://github.com/openlayers/openlayers/pull/9259 .

**Comment by icholy** [comment]
@JulioJu this is not the right place for this discussion. Please use Stack Overflow, or at the very least, open a separate issue.

**Comment by JulioJu** [comment]
Sorry, very sorry probably I must delete my previous comments ? I spam and confuse the current issue. Very sorry.

**Comment by bampakoa** [comment]
@ahocevar Good news 🎉 `--declaration` and `--allowJs` are now in Typescript 3.7 Beta 😄
